pkg/inlinekeepalive/client.go

```diff
@@ -0,0 +1,119 @@
+package inlinekeepalive
+
+import (
+	"context"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+)
+
+// KeepaliveClientStream implements grpc.ClientStream
+type KeepaliveClientStream struct {
+	log     hclog.Logger
+	handler grpc.ClientStream
+}
+
+func (k *KeepaliveClientStream) Header() (metadata.MD, error) {
+	return k.handler.Header()
+}
+
+func (k *KeepaliveClientStream) Trailer() metadata.MD {
+	return k.handler.Trailer()
+}
+
+func (k *KeepaliveClientStream) CloseSend() error {
+	return k.handler.CloseSend()
+}
+
+func (k *KeepaliveClientStream) Context() context.Context {
+	return k.handler.Context()
+}
+
+func (k *KeepaliveClientStream) SendMsg(m interface{}) error {
+	return k.handler.SendMsg(m)
+}
+
+// RecvMsg intercepts keepalive messages and does not pass them
+// along to the handler.
+func (k *KeepaliveClientStream) RecvMsg(m interface{}) error {
+	for {
+		err := k.handler.RecvMsg(m)
+		if err != nil {
+			return err
+		}
+
+		pm, ok := m.(protoreflect.ProtoMessage)
+		if !ok {
+			// Weird, not a protobuf message, but not our keepalive, so continue as normal
+			return nil
+		}
+
+		if !IsInlineKeepalive(k.log, pm) {
+			return nil
+		}
+
+		// It's a keepalive message! Ignore it and recv again
+		continue
+	}
+}
+
+// isServerCompatible determines if a server is able to receive inline keepalives
+// by examining the features it advertises.
+func isServerCompatible(ctx context.Context, cc *grpc.ClientConn) (bool, error) {
+	client := pb.NewWaypointClient(cc)
+
+	versionInfo, err := client.GetVersionInfo(ctx, &emptypb.Empty{})
+	if err != nil {
+		return false, errors.Wrapf(err, "failed getting version info to determine if server is inline-keepalive compatible")
+	}
+
+	for _, feature := range versionInfo.Features {
+		if feature == FeatureName {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// KeepaliveClientStreamInterceptor returns a stream interceptor
+// that sends inline keepalive messages on client streams (if the server
+// is compatible), and intercepts inline keepalives from the server.
+func KeepaliveClientStreamInterceptor() grpc.StreamClientInterceptor {
+	return func(
+		ctx context.Context,
+		desc *grpc.StreamDesc,
+		cc *grpc.ClientConn,
+		method string,
+		streamer grpc.Streamer,
+		opts ...grpc.CallOption,
+	) (grpc.ClientStream, error) {
+		log := hclog.FromContext(ctx).With("method", method)
+
+		ctx = metadata.AppendToOutgoingContext(ctx, HeaderSendKeepalives, "true")
+
+		handler, err := streamer(ctx, desc, cc, method, opts...)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get the client handler when setting up the inlinekeepalive interceptor on method %q", method)
+		}
+
+		serverCompatible, err := isServerCompatible(ctx, cc)
```

Comment:
Is the intercepter called on every method? If so, this code is checking the version on every RPC rather than just once per clientConn.

Reply:
Yup, it's calling GetVersionInfo once per client-streaming RPC. I think that's the safest thing to do - i'd rather not assume that each conn is pointing at the same waypoint server, and I also don't have a great spot to put the fact that we've already checked the server's features. I added this call to the goroutine, so it won't block main body of streaming logic.

Added a bundle of new comments, added to doc.go, and added a README with diagrams spelling this out.

Let me know if there's anywhere else you think I should add comments.